design/unnumbered-bgp.md

```diff
@@ -0,0 +1,498 @@
+// == extra FRR directive `neighbor fe80::a876:4dff:fe77:408 interface eth0`
+// 4. If address is LLA and empty interface, then fail
+// 5. If address is empty and interface has value then the BGP unnumber peering config will take place
+// ==  modify FRR directive `neighbor PEER interface remote-as <ASN>`
+// && replace any address directive to use interface e.g.  `neighbor net0 timers 180 540`
+// && enable RAs on this interface
+// 6. If address has value (non-LLA) and interface non empty, then fail
+```
+That option might be not ideal (subjective opinion) because documentation looks complicated,
+and user might have to do trial-and-error approach. Plus user needs to understand corner cases.
+
+##### FRR-K8S
+
+```go
+...
+
+type Neighbor struct {
+  Address string `json:"address"`
+  Interface string `json:"interface"`
+}
+
+```
+
+##### MetalLB
+
+```go
+type BGPPeerSpec struct {
+  Address string `json:"peerAddress"`
+  Interface bool `json:"interface"`
+}
+
+```
+
+Example CR
+
+```yaml
+apiVersion: metallb.io/v1beta2
+kind: BGPPeer
+metadata:
+  name: peer-unnumber
+  namespace: metallb-system
+spec:
+  myASN: 64512
+  ASN: 645000
+  peerAddress: "" // empty, can be omitted
+  interface: net0
+```
+
+#### Option 3. Add new Unnumber bool field
+
+We can introduce a new field `Unnumber bool`, which if true the we indicate
+Unnumber BGP Peering. If `true` the Address field holds the interface value.
+Bool option makes the configuration error-free, because the user declares
+explicitly that they want Unnumber and there is no "guessing" what metallb
+does. Unnumber case is disconnected from LLA support case.
+
+That option might be not ideal (subjective opinion) because the `Address` field
+holds a value that is interface (ideally Address should be called Peer), having
+a field changing its functionality according to a different field seems
+un-intuitive. E.g. the value in `Address` will add those lines
+
+```
+interface net0
+  ipv6 nd ra-interval 10
+  no ipv6 nd suppress-ra
+```
+
+##### FRR-K8S
+```go
+
+type Neighbor struct {
+  Address string `json:"address"` // valid IP or interface if unnumber is true
+  Unnumber bool `json:"Unnumber"`
+}
+```
+
+##### MetalLB
+
+```go
+type BGPPeerSpec struct {
+  Address string `json:"peerAddress"`
+  Unnumber bool `json:"unnumber"`
+}
+```
+
+Example CR
+
+```yaml
+apiVersion: metallb.io/v1beta2
+kind: BGPPeer
+metadata:
+  name: peer-unnumber
+  namespace: metallb-system
+spec:
+  myASN: 64512
+  ASN: 645000
+  peerAddress: net0
+  unnumber: true
+```
+
+### Testing
+
+#### E2E tests
+
+- Modify the testing infra to include point-to-point links which means connect
+containers using VETH, and not through a bridge. We need NET-ADMIN cap (or sudo)
+when creating the VETH.
+
+- Create/Delete peer per that specific test not to take resources from runners
+outside the runtime of the test. This peer needs different configuration (not
+IPs and interface names), keep separate raw configuration.
```

Comment:
not sure I understand, you're proposing a separate test for the unnumbered case or expanding the existing containers setup?

Reply:
I propose a separate isolated test for that use case.
(but no strong argument against expanding container setup).
Do you have a preference here?

Reply:
I'd say if it's possible to have all the tests ran on unnumbered in a clean way it's preferable